### PR TITLE
deepspeed/launcher: add launcher_helper as each rank's start portal

### PR DESCRIPTION
**Pull request by YizhouZ**
File Changes:
multinode_runner.py: modify mpich runner to use launcher_helper
launcher_helper.py: init script to map env variables

Descriptions:
Previous mpich runner would cause linux command line reaching size limitations when rank number is extremely higher. After  discussion, we want to optimize it by using a helper script as each rank's start portal, which maps env variables such as rank, local_rank for deepspeed. So far we only use it for mpich runner, but it is made to be extendable, any runner could be added if facing similar situation. Only necessary args are passed to helper script.

Let us know if there is any suggestion.

### CODE REVIEW

**Comment by YizhouZ** [comment]
@tjruwase Hi, could you please help to start workflow for this PR? Thanks.

**Comment by YizhouZ** [comment]
Hi @jeffra @awan-10 @tjruwase, is there any comment on this PR? Thanks!

**Comment by delock** [comment]
Hi @YizhouZ can you show the command line launched by deepspeed before and after your PR, illuatrating how your PR could help reduce command line length?  Thanks!

**Comment by YizhouZ** [comment]
> Hi @YizhouZ can you show the command line launched by deepspeed before and after your PR, illuatrating how your PR could help reduce command line length? Thanks!

Sure.

Without this PR, while using mpich runner, the cmd would be like
```
cmd = mpirun --genv xxx -n 1 -env xxx python xxx.py : -n 1 -env xxx python xxx.py : <repeat local number times> : -n 1 -env xxx python xxx.py
``` 
When the number of total ranks goes high, the cmd would be extremely long, and trigger cmd word size limitations.

After this PR, the cmd would be much shorter:
```
cmd = mpirun --genv xxx python xxx.py <with no more additional cmds>
```

**Comment by delock** [comment]
Hi @mrwyattii Do you have any comments on this PR?  This PR is essential when need to run DeepSpeed training on thousands of nodes.  The former implementation would make command line too long that overflow the command line buffer.  The new implementation fix this issue.

**Comment by YizhouZ** [comment]
Hi @tjruwase @mrwyattii, do you have any comments on this PR? Like @delock said previously, this new implement is essential for us to enable DeepSpeed Training on a large amount of nodes. Otherwise training process would reach linux command limits.

**Comment by tjruwase** [comment]
@YizhouZ, @delock, thanks for this great PR. Apologies for the delay. We will review asap.

File: deepspeed/launcher/multinode_runner.py [hunk 1]
```diff
@@ -210,32 +214,29 @@ def get_cmd(self, environment, active_resources):
...
+            if self.args.module:
+                python_exec.append("-m")
+                helper_args.append("--module")
+        else:
+            helper_args.append("--no_python")
+
+        helper_cmd = str(os.path.dirname(os.path.realpath(__file__))) + '/launcher_helper.py'
+        herlper_cmd = [helper_cmd] + helper_args + [self.user_script] + self.user_arguments
+
+        return mpirun_cmd + export_cmd + python_exec + herlper_cmd
```
Comment (mrwyattii) [inline deepspeed/launcher/multinode_runner.py:239]:
```suggestion
        helper_cmd = [helper_cmd] + helper_args + [self.user_script] + self.user_arguments

        return mpirun_cmd + export_cmd + python_exec + helper_cmd
```

**Review by mrwyattii — APPROVED** [review]

**Review by loadams — APPROVED** [review]

**Comment by YizhouZ** [comment]
Thank you so much for this quick merge! @tjruwase @mrwyattii @loadams

**Pull request merged**